### DISCUSSION TITLE
[Craft 4] Filesystems & Volumes

### DISCUSSION THREAD

**Discussion by heymarkreeves** [craftcms/cms#11044] [Ideas]
I started working with Craft 4 RC1 yesterday, including setting up filesystems & volumes.

I'd be interested in the possibility of 1:many filesystem:volumes, so that I could set up a filesystem to an S3 or Digital Ocean Spaces **bucket once**, and then set up **many volumes** that specify **folders** in that bucket. I didn't see an option to specify folders when creating a volume.

Use case: for the scope of the sites I typically build, all of the images just go into 1 site-specific bucket, but I create folders for each type and restrict Assets fields to only relevant volumes. On Craft 3, I'd set up volumes for all the different types of images: heroes, landscape, user photos, icons, etc, to keep them organized in this way. In Craft 4, it seems I need both a filesystem **and** a volume for each type of image, where those are really just distinct volumes using one filesystem.

**Comment by richhayler** [comment]
@heymarkreeves i had the same thought. It seems that filesystems and volumes have a 1 to 1 relationship, i.e. once you select a filesystem in a volume’s settings that file system doesn’t appear for any new volumes. It seems that filesystems to volumes should be 1 to many. 

I agree with you that allowing the same filesystem to be selectable for any number of volumes and in the volume settings have an option to specify a folder (a subfolder within the parent filesystem) to store files uploaded against that volume.

**Reply by richhayler** [reply]
@heymarkreeves is this something along the lines of what you were thinking (added the section highlighted in red)? 

![CleanShot 2022-05-13 at 15 33 51@2x](https://user-images.githubusercontent.com/2505643/168218252-54c20142-6866-4fc7-8462-e640502ab166.png)

**Comment by heymarkreeves** [comment]
@richhayler Yes, and you'd need the ability to choose the same filesystem for multiple volumes. So I create the filesystem and choose 1 S3 bucket there, and then I create _n_ volumes that use that filesystem, with each volume specifying a different subfolder as their root.

**Comment by mylesthedev** [comment]
I just setup a new 4 site and had to check if this was a bug... In my head a filesystem should have many volumes. Seems quite messy to have multiple FS e.g. common is images and videos
